enhancements/network/on-prem-service-load-balancers.md

```diff
@@ -0,0 +1,405 @@
+  cluster.
+* Make our SLB controller smart enough to understand different IP address
+  pools, their associated L2 domains, and which Nodes are on which L2 domain.
+* Extend keepalived (either directly or via some integration) to support an L3
+  based address location advertisement (likely BGP).
+
+Something based on keepalived is probably our simplest solution.  However,
+downsides include:
+
+* This would be entirely built by us.  It’s possible we could build some
+  community usage around a simple solution like this, but that would take time.
+  Given more featureful alternatives that already exist, I wouldn't expect much
+  traction.
+* There’s not a lot of opportunity for future functionality growth here unless
+  we start swapping out the pieces (keepalived and/or haproxy), which would
+  also sacrifice some of the simplicity.
+* Keepalived uses the VRRP protocol and it would be nice to move away from
+  this.  VRRP IDs are just 0-255 and the Keepalived + haproxy integration for
+  bare metal IPI generates VRRP IDs based on the cluster name.  Even with
+  different names, it’s possible to have a collision, and that can cause
+  problems in lab environments with a lot of test clusters on shared networks.
+  VRRP uses multicast by default which is not allowed in all environments,
+  though it's also possible to configure keepalived to use unicast, instead.
+
+### kube-vip
+
+* [Web site](https://kube-vip.io/)
+* [Kube-vip docs for SLBs](https://kube-vip.io/kubernetes/)
+* [Code](https://github.com/plunder-app/kube-vip)
+
+I came across `kube-vip` when the author shared it in the
+`#cluster-api-provider` channel on the Kubernetes slack.  It’s new and likely
+not mature, but some of the implementation is clever.
+
+Instead of using VRRP to provide IP address HA, it uses RAFT (from
+[https://github.com/hashicorp/raft]).  That would help avoid potential VRRP ID
+collisions between multiple clusters.
+
+Kube-vip implements its own custom load balancer, which is concerning from a
+security, feature, and performance perspective.
+
+`kube-vip` uses a couple of other supporting components:
+* [starboard](https://github.com/plunder-app/starboard) - daemonset, manages
+  iptables rules based on current IP address location
+* [plndr-cloud-provider](https://github.com/plunder-app/plndr-cloud-provider) -
+  kubernetes cloud provider
+
+Kube-vip only supports layer 2 based address advertisement, and it doesn’t look
+like it supports IPv6 yet.
+
+Despite the project's young age, someone has already [integrated it with
+OpenShift 3.11](https://github.com/megian/openshift-kube-vip-ansible).
+
+Some of the key downsides to this option:
+* Depends on yet-another-raft-implementation --
+  https://github.com/hashicorp/raft
+* New, developed as a hobby project by one person, likely PoC level maturity
+* Lacking any layer 3 based address advertisement options
+
+### OVN-Kubernetes Native Solution
```

Comment:
This is an interesting proposal, and I agree it's probably not the right solution right now, but it's definitely interesting as  a motivating user story for the future. FYI @dcbw.

Reply:
Thanks for that insight, @aojea.  At least with ovn-kubernetes there's also a central controller component that could take on the LoadBalancer management, so it wouldn't have to go into the CNI plugin directly.

Reply:
> ovn-kubernetes there's also a central controller component that could take on the LoadBalancer management,

my point is that Load Balancers have different lifecycles and features, the release process and development is more complicated if you tie it to the CNI controller ... we made that mistake with Midonet too :sweat_smile:  @celebdor may remember those funny days :wink: and I think that Openstack did kind of the same with the neutron-lbaas project -> Octavia, but maybe wrong here and the reasons were others.
Let's assume you receive a new feature for the load balancer or a bug fix, it forces you to cut a new release for the ovn-kubernetes ... same happen once you want to roll back a new version or with an upgrade. You'll have different users, roadmaps and problems in the same code base :shrug: